tensorflow/contrib/layers/kernels/layer_norm_fused_op_gpu.cu.cc

```diff
@@ -0,0 +1,1455 @@
+                                               const int slice_per_block) {
+  const int slice_size = args.slice_size;
+  const int in_depth = args.depth;
+  const int n_inputs = args.n_inputs;
+  const T epsilon = args.epsilon;
+
+  const T i_n = static_cast<T>(1.0f) / static_cast<T>(in_depth);
+
+  const int slice_id = threadIdx.x / slice_size;
+  const int thread_slice_id = threadIdx.x % slice_size;
+
+  const T _beta =
+      get_value<T>(beta + thread_slice_id, thread_slice_id, in_depth);
+
+  T mu;
+  T rstd;
+
+  for (int bId = blockIdx.x; bId < num_blocks; bId += gridDim.x) {
+    mu = static_cast<T>(0.0f);
+    rstd = static_cast<T>(0.0f);
+
+    const int thread_id =
+        (bId * slice_per_block + slice_id) * in_depth + thread_slice_id;
+    // const T inp = 0;
+    const T inp = get_value<T>(input + thread_id, thread_slice_id, in_depth);
+
+    mu += inp * i_n;
+
+    for (int mask = slice_size / 2; mask > 0; mask /= 2) {
+      mu += __shfl_xor(mu, mask);
+    }
+
+    if (thread_slice_id < in_depth) rstd += (inp - mu) * (inp - mu);
+
+    for (int mask = slice_size / 2; mask > 0; mask /= 2) {
+      rstd += __shfl_xor(rstd, mask);
+    }
+
+    rstd = rsqrt(rstd * i_n + epsilon);
+
+    if (thread_slice_id < in_depth && thread_id < n_inputs)
+      output[thread_id] = (inp - mu) * rstd + _beta;
+  }
+}
+
+template <typename T, int mult>
+__global__ void LayerNormFusedGPUKernel(const LayerNormFusedArgs args,
+                                        const T* __restrict__ input,
+                                        const T* __restrict__ gamma,
+                                        const T* __restrict__ beta,
+                                        T* __restrict__ output,
+                                        const int num_blocks) {
+  const int in_depth = args.depth;
+  const int slice_size = args.slice_size;
+  const int n_inputs = args.n_inputs;
+  const T epsilon = args.epsilon;
+
+  const int thread_warp_id = threadIdx.x % warpSize;
+
+  const int thread_slice_id = threadIdx.x % slice_size;
+
+  extern __shared__ __align__(sizeof(T)) unsigned char my_smem[];
+  T* mean_cache = (T*)my_smem;
+  T* std_cache = (T*)&my_smem[sizeof(T)];
+
+  const T i_n = static_cast<T>(1.0f) / static_cast<T>(in_depth);
+  T inp[mult];
+  T _gamma[mult];
+  T _beta[mult];
+  int thread_id[mult];
+
+  T sum;
+  T sqSum;
+  T mu;
+  T rstd;
+
+  UNROLL for (int m = 0; m < mult; m++) {
+    _gamma[m] = get_value<T>(gamma + thread_slice_id + m * blockDim.x,
+                             thread_slice_id + m * blockDim.x, in_depth);
+    _beta[m] = get_value<T>(beta + thread_slice_id + m * blockDim.x,
+                            thread_slice_id + m * blockDim.x, in_depth);
+  }
+
+  for (int bId = blockIdx.x; bId < num_blocks; bId += gridDim.x) {
+    sum = static_cast<T>(0.0f);
+    sqSum = static_cast<T>(0.0f);
+
+    if (thread_slice_id == 0) {
+      mean_cache[0] = static_cast<T>(0.0f);
```

Comment:
Since this is not really an array, I feel it is more clear to say "*mean_cache", instead of "mean_cache[0]".